fem/quadinterpolator.hpp

```diff
@@ -0,0 +1,222 @@
+   void Mult(const Vector &e_vec, unsigned eval_flags,
+             Vector &q_val, Vector &q_der, Vector &q_det) const;
+
+   /// Perform the transpose operation of Mult(). (TODO)
+   void MultTranspose(unsigned eval_flags, const Vector &q_val,
+                      const Vector &q_der, Vector &e_vec) const;
+
+   // Compute kernels follow (cannot be private or protected with nvcc)
+
+   /// Template compute kernel for 2D.
+   template<const int T_VDIM = 0, const int T_ND = 0, const int T_NQ = 0>
+   static void Eval2D(const int NE,
+                      const int vdim,
+                      const DofToQuad &maps,
+                      const Vector &e_vec,
+                      Vector &q_val,
+                      Vector &q_der,
+                      Vector &q_det,
+                      const int eval_flags);
+
+   /// Template compute kernel for 3D.
+   template<const int T_VDIM = 0, const int T_ND = 0, const int T_NQ = 0>
+   static void Eval3D(const int NE,
+                      const int vdim,
+                      const DofToQuad &maps,
+                      const Vector &e_vec,
+                      Vector &q_val,
+                      Vector &q_der,
+                      Vector &q_det,
+                      const int eval_flags);
+};
+
+/** @brief A class that performs interpolation from an E-vector to quadrature
+    point values and/or derivatives (Q-vectors) on the faces. */
+/** An E-vector represents the element-wise discontinuous version of the FE
+    space and can be obtained, for example, from a GridFunction using the
+    Operator returned by FiniteElementSpace::GetElementRestriction().
+
+    The target quadrature points in the elements can be described either by an
+    IntegrationRule (all mesh elements must be of the same type in this case) or
+    by a QuadratureSpace. */
+class FaceQuadratureInterpolator
+{
+private:
+   FaceType type;
+   int nf;
+protected:
+   friend class FiniteElementSpace; // Needs access to qspace and IntRule
+
+   const FiniteElementSpace *fespace;  ///< Not owned
+   const QuadratureSpace *qspace;      ///< Not owned
+   const IntegrationRule *IntRule;     ///< Not owned
+
+   mutable bool use_tensor_products;
+
+   static const int MAX_NQ1D = 10;
+   static const int MAX_ND1D = 10;
+   static const int MAX_VDIM1D = 1;
+
+   static const int MAX_NQ2D = 100;
+   static const int MAX_ND2D = 100;
+   static const int MAX_VDIM2D = 2;
+
+   static const int MAX_NQ3D = 1000;
+   static const int MAX_ND3D = 1000;
+   static const int MAX_VDIM3D = 3;
+
+public:
+   enum FaceEvalFlags
+   {
+      VALUES       = 1 << 0,  ///< Evaluate the values at quadrature points
+      DERIVATIVES  = 1 << 1,  ///< Evaluate the derivatives at quadrature points
+      /** @brief Assuming the derivative at quadrature points form a matrix,
+          this flag can be used to compute and store their determinants. This
+          flag can only be used in Mult(). */
+      DETERMINANTS = 1 << 2,
```

Comment:
`DETERMINANTS` is not quite accurate here (the Jacobians are not square), may call it `WEIGHTS`?